mouseion/mouseion/train_and_build_index.py

```diff
@@ -0,0 +1,82 @@
+import argparse
+import faiss
+import json
+import numpy
+import tqdm
+import pathlib
+
+def main(args):
+    """
+    Combines numpy embeddings from multiple npy files into a FAISS index.
+    """
+    try:
+        spec = json.load(args.spec.open("r"))
+    except Exception as e:
+        print(f"Error reading spec file: {e}")
+        return
+
+    total_embeddings = sum([numpy.load(shard["embeddings"], mmap_mode="r").shape[0] for shard in spec["shards"]])
+    is_trained = False
+
+    try:
+        first_shard = numpy.load(spec["shards"][0]["embeddings"], mmap_mode="r")
+    except Exception as e:
+        print(f"Error loading shard: {e}")
+        return
+
+    dim = first_shard.shape[1]
+    print(f'dimensionality is {dim}')
+
+    index = faiss.index_factory(dim, args.index_type, faiss.METRIC_L2)
+
+    if args.use_gpus:
+        gpu_resources = []
+        for i in range(faiss.get_num_gpus()):
+            res = faiss.StandardGpuResources()
+            res.setTempMemory(2**20)
+            gpu_resources.append(res)
+
+        co = faiss.GpuMultipleClonerOptions()
+        co.useFloat16 = True
+        co.usePrecomputed = True
+        co.shard = True
+        co.resources = gpu_resources
+
+        index = faiss.index_cpu_to_all_gpus(index, co)
+
+    with tqdm.tqdm(total=total_embeddings, desc="Embeddings") as progress:
```

Comment:
While it may be less pretty/exact, I would consider handing `spec["shards"]` directly to `tqdm.tqdm()` for cleaner iteration.

Reply:
would you suggest the corrected tqdm line? I am barely familiar with tqdm, and not clear what you had in mind.

We need to keep batching intact (the loop below) or OOM ensues.